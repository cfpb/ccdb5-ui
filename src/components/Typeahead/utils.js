/**
 * Sets the options for the async typeahead
 *
 * @param {string} value - value typed into the typeahead
 * @param {Function} setOptions - function to set local array of options
 * @param {string} uri - the endpoint being hit by the fetch
 * @returns {void | Promise<object>} - Either nothing or a response with option object
 */
export function handleFetchSearch(value, setOptions, uri) {
  const num = value.toLowerCase();
  if (num === '') {
    setOptions([]);
    return;
  }
  return fetch(uri)
    .then((result) => result.json())
    .then((items) => {
      const options = items.map((item) => ({
        key: item,
        label: item,
        position: item.toLowerCase().indexOf(num),
        value,
      }));
      setOptions(options);
    });
}
