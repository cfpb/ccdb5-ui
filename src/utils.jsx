import { DATE_RANGE_MIN, SLUG_SEPARATOR } from './constants'
import moment from 'moment'

// eslint-disable-next-line complexity
export const calculateDateInterval = ( minDate, maxDate ) => {
  // only check intervals if the end date is today
  // round off the date so the partial times don't mess up calculations
  const today = moment( new Date() ).startOf( 'day' )
  const end = moment( maxDate ).startOf( 'day' )
  const start = moment( minDate ).startOf( 'day' )

  // make sure end date is the same as today's date
  if ( end.diff( today, 'days', true ) !== 0 ) {
    return ''
  }

  // is the start date the same as the oldest document?
  if ( moment( minDate ).isSame( DATE_RANGE_MIN, 'day' ) ) {
    return 'All'
  }

  // verify if it's 3 or 1 years
  const yrDiff = end.diff( start, 'years', true )

  if ( yrDiff === 3 || yrDiff === 1 ) {
    return yrDiff + 'y'
  }

  // verify if it's 6 or 3 months
  const moDiff = end.diff( start, 'months', true )
  if ( moDiff === 6 || moDiff === 3 ) {
    return moDiff + 'm'
  }

  return ''
}

/**
 * Function to set the limit of the range of a set of numbers
 * @param {int} x value we are checking
 * @param {int} min smallest number it can me
 * @param {int} max biggest number it can be
 * @returns {*}the limited value
 */
export const clamp = ( x, min, max ) => {
  if ( x < min ) {
    x = min;
  } else if ( x > max ) {
    x = max;
  }
  return x;
}

/**
 * Replacement for the common pattern:
 * if( o.field )
 *    x = o.field
 * else
 *    x = alternateValue
 *
 * Avoids some of the complexity lint warnings
 *
 * @param {Object} o the object being tested
 * @param {string} field the field to check
 * @param {string|Object} alternateValue the value to use in absence
 * @returns {string} the value to use
 */
export const coalesce = ( o, field, alternateValue ) => {
  if ( typeof o !== 'object' ) {
    return alternateValue;
  }

  return field in o && o[field] ? o[field] : alternateValue;
};

/**
 * Creates a hash from a string
 *
 * @param {string} someString the string to hash
 * @returns {int} a hashing of the string
 */
export function hashCode( someString ) {
  const s = String( someString )
  let hash = 0
  let i, chr
  if ( s.length === 0 ) {
    return hash
  }
  for ( i = 0; i < s.length; i++ ) {
    chr = s.charCodeAt( i )
    hash = ( hash << 5 ) - hash + chr

    // Convert to 32bit integer
    hash |= 0
  }
  return hash
}

/**
 * Creates a hash from an object
 *
 * @param {string} o the object to hash
 * @returns {string} a hashing of the object
 */
export function hashObject( o ) {
  return hashCode( JSON.stringify( o ) )
}
export const normalize = s => s.toLowerCase()

export const slugify = ( a, b ) => a + SLUG_SEPARATOR + b

export const sortSelThenCount = ( options, selected ) => {
  const retVal = ( options || [] ).slice()

  // Sort the array so that selected items appear first, then by doc_count
  /* eslint complexity: ["error", 5] */
  retVal.sort( ( a, b ) => {
    const aSel = selected.indexOf( a.key ) !== -1
    const bSel = selected.indexOf( b.key ) !== -1

    if ( aSel && !bSel ) {
      return -1;
    }
    if ( !aSel && bSel ) {
      return 1;
    }

    // Both are selected or not selected
    // Sort by descending doc_count
    return b.doc_count - a.doc_count
  } )

  return retVal
}

/**
* Safely format a date
*
* @param {Date} date the date to convert
* @returns {string} the date formatted for the current locale
*/
export function shortFormat( date ) {
  const wrapped = moment( date )
  return date ? wrapped.format( 'M/D/YYYY' ) : ''
}

/**
* Convert a date to a truncated ISO-8601 string
*
* @param {Date} date the date to convert
* @returns {string} the date formatted as yyyy-mm-ddd
*/
export function shortIsoFormat( date ) {
  return date ? date.toISOString().substring( 0, 10 ) : ''
}

// ----------------------------------------------------------------------------
// attribution: underscore.js (MIT License)

/**
* Delay the implementation of a function until after a period of time
* This prevents expensive calls from being made while triggering events are
* still happening
*
* @param {function} func a function with an embedded expensive call
* @param {int} wait the number of msecs to delay before calling the function
* @returns {function} a replacement function to use in place of the original
*/
export function debounce( func, wait ) {
  var timer = null;

  var later = ( context, args ) => {
    timer = null;
    func.apply( context, args );
  }

  return ( ...theArgs ) => {
    if ( !timer ) {
      timer = setTimeout( later, wait, this, theArgs )
    }
  }
}

// ----------------------------------------------------------------------------
// attribution: lodash.js (Creative Commons License)

/**
* Binds methods of an object to the object itself, overwriting the existing
* method
*
* @param {Object} obj The object to bind and assign the bound methods to.
* @param {...(string|string[])} methodNames The object method names to bind,
*  specified individually or in arrays.
* @returns {Object} the updated object
*/
export function bindAll( obj, methodNames ) {
  const length = methodNames.length
  for ( let i = 0; i < length; i++ ) {
    const key = methodNames[i]
    obj[key] = obj[key].bind( obj )
  }
  return obj;
}
