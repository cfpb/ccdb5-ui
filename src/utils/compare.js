export const arrayEquals = (first, second) => {
  let length = first.length;
  if (length !== second.length) return false;
  while (length--) {
    if (first[length] !== second[length]) return false;
  }
  return true;
};

/**
 * Function to compare two objects or arrays.
 *
 * @param {object | Array | any} first - value 1 to convert to string
 * @param {object | Array | any} second - value 2 to convert to string to compare
 * @returns {boolean} Whether the values are equal
 */
export const isEqual = (first, second) =>
  JSON.stringify(first) === JSON.stringify(second);
