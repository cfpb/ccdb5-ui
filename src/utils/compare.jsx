export const arrayEquals = (a, b) => {
  let i = a.length;
  if (i !== b.length) return false;
  while (i--) {
    if (a[i] !== b[i]) return false;
  }
  return true;
};
