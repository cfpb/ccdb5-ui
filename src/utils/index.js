import {
  DATE_RANGE_MIN,
  flagFilters,
  knownFilters,
  SLUG_SEPARATOR,
} from '../constants/index';
import Analytics from '../actions/analytics';
import dayjs from 'dayjs';
import queryString from 'query-string';

/**
 * Breaks up '123' to '1 2 3' to help screen readers read digits individually
 * https://thatdevgirl.com/blog/accessibility-phone-number-formatting
 *
 * @param {string} digits - the string of digits
 * @returns {string} an expanded string of digits
 */
export function ariaReadoutNumbers(digits) {
  return Array.from(digits || '').join(' ');
}

/* eslint-disable no-console */

// eslint-disable-next-line complexity
export const calculateDateRange = (minDate, maxDate) => {
  // only check intervals if the end date is today
  // round off the date so the partial times don't mess up calculations
  const today = dayjs(startOfToday());
  const end = dayjs(maxDate).startOf('day');
  const start = dayjs(minDate).startOf('day');

  // make sure end date is the same as today's date
  if (end.diff(today, 'days') !== 0) {
    return '';
  }

  // is the start date the same as the oldest document?
  if (dayjs(minDate).isSame(DATE_RANGE_MIN, 'day')) {
    return 'All';
  }

  // verify if it's 3 or 1 years
  const yrDiff = end.diff(start, 'years', true);
  if (yrDiff === 3 || yrDiff === 1) {
    return yrDiff + 'y';
  }

  // verify if it's 6 or 3 months
  const moDiff = end.diff(start, 'months', true);
  if (moDiff === 6 || moDiff === 3) {
    return moDiff + 'm';
  }

  return '';
};

/**
 * Clamps number within the inclusive lower and upper bounds.
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_clamp
 *
 * @param {number} number - The value we are checking.
 * @param {number} boundOne - The lower bound we don't want to go before.
 * @param {number} boundTwo - The upper bound we can't go past.
 * @returns {number} The clamped number.
 */
export const clamp = (number, boundOne, boundTwo) => {
  if (!boundTwo) {
    return Math.max(number, boundOne) === boundOne ? boundOne : number;
  } else if (Math.min(number, boundOne) === number) {
    return boundOne;
  } else if (Math.max(number, boundTwo) === number) {
    return boundTwo;
  }
  return number;
};

/**
 * Function to set the limit of the range of a set of dates
 *
 * @param {string} val - value we are checking
 * @param {string} min - smallest number it can be
 * @param {string} max - biggest number it can be
 * @returns {*} the limited value
 */
export const clampDate = (val, min, max) => {
  let xDate = new Date(val);
  const minDate = new Date(min);
  const maxDate = new Date(max);

  if (xDate < minDate) {
    xDate = minDate;
  } else if (xDate > maxDate) {
    xDate = maxDate;
  }
  return xDate;
};

/**
 * Replacement for the common pattern:
 * if( o.field )
 * x = o.field
 * else
 * x = alternateValue
 *
 * Avoids some of the complexity lint warnings
 *
 * @param {object} object - the object being tested
 * @param {string} field - the field to check
 * @param {string | object} alternateValue - the value to use in absence
 * @returns {string | Array | object} the value to use
 */
export const coalesce = (object, field, alternateValue) => {
  if (typeof object !== 'object') {
    return alternateValue;
  }

  return field in object && object[field] ? object[field] : alternateValue;
};

/**
 * Creates a hash from a string
 *
 * @param {string} someString - the string to hash
 * @returns {number} a hashing of the string
 */
export function hashCode(someString) {
  const str = String(someString);
  let hash = 0;
  let index, chr;
  if (str.length === 0) {
    return hash;
  }
  for (index = 0; index < str.length; index++) {
    chr = str.charCodeAt(index);
    hash = (hash << 5) - hash + chr;

    // Convert to 32bit integer
    hash |= 0;
  }
  return hash;
}

/**
 * helper function to determine if we have any filters selected so we can
 * disable the Per 1000 Complaints button
 * enable per1000 if the only filter selected is state
 *
 * @param {object} filters - reducer contains values for the filters, etc
 * @returns {boolean} are we enabling the perCap
 */
// eslint-disable-next-line complexity
export function enablePer1000(filters) {
  const keys = [];
  let filter;
  const allFilters = knownFilters.concat(flagFilters);

  for (let index = 0; index < allFilters.length; index++) {
    filter = allFilters[index];

    if (
      (Array.isArray(filters[filter]) && filters[filter].length) ||
      filters[filter] === true
    ) {
      keys.push(filter);
    }
  }
  const compReceivedFilters = ['company_received_max', 'company_received_min'];
  for (let index = 0; index < compReceivedFilters.length; index++) {
    filter = compReceivedFilters[index];
    if (filters[filter]) {
      keys.push(filter);
    }
  }

  if (keys.length) {
    // normalization still okay as long as only State filters are applied
    return keys.length === 1 && keys[0] === 'state';
  }

  return true;
}

export const normalize = (str) => str.toLowerCase();

/**
 * takes a string and formats it into proper text for an htmd ID
 * Eat at Joe's => eatatjoes
 *
 * @param {string} str - the dirty string Eat at Joe's
 * @returns {string} sanitized string eat-at-joe-s
 */
export const sanitizeHtmlId = (str) =>
  str.replace(/\s+|\W/g, '-').toLowerCase();

export const slugify = (first, second) => first + SLUG_SEPARATOR + second;

/**
 * Custom sort for array so that selected items appear first, then by doc_count
 *
 * @param {Array} options - input array containing values
 * @param {Array} selected - values
 * @returns {Array} sorted array
 */
export const sortSelThenCount = (options, selected) => {
  const retVal = (structuredClone(options) || []).slice();

  /* eslint complexity: ["error", 5] */
  retVal.sort((first, second) => {
    const aSel = selected.indexOf(first.key) !== -1;
    const bSel = selected.indexOf(second.key) !== -1;

    if (aSel && !bSel) {
      return -1;
    }
    if (!aSel && bSel) {
      return 1;
    }

    // Both are selected or not selected
    // Sort by descending doc_count
    return second.doc_count - first.doc_count;
  });

  return retVal;
};

/**
 * Safely format a date
 *
 * @param {Date} date - the date to convert
 * @returns {string} the date formatted for the current locale
 */
export function shortFormat(date) {
  const wrapped = dayjs(date);
  return date ? wrapped.format('M/D/YYYY') : '';
}

/**
 * Convert a date to a truncated ISO-8601 string
 *
 * @param {string | object | Date} date - the date to convert
 * @returns {string} the date formatted as yyyy-mm-ddd
 */
export function shortIsoFormat(date) {
  if (typeof date === 'string') {
    return date.slice(0, 10);
  } else if (typeof date === 'object' && date !== null) {
    return dayjs(date).toISOString().slice(0, 10);
  }
  return '';
}

/**
 * Gets the UTC time for the beginning of the day in the local time zone
 *
 * @returns {string} midnight today, local
 */
export function startOfToday() {
  if (!Object.prototype.hasOwnProperty.call(window, 'MAX_DATE')) {
    if (
      Object.prototype.hasOwnProperty.call(window, 'complaint_public_metadata')
    ) {
      const { metadata_timestamp: stamp } = window.complaint_public_metadata;
      window.MAX_DATE = new Date(dayjs(stamp).startOf('day').toString());
    } else {
      console.error('complaint_public_metadata is missing');
      window.MAX_DATE = new Date(dayjs().startOf('day').toString());
    }
  }

  // Always return a clone so the global is not exposed or changed
  return new Date(window.MAX_DATE.valueOf());
}

// ----------------------------------------------------------------------------
// attribution: underscore.js (MIT License)

/**
 * Native implementation of lodash debounce
 * https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore#_debounce
 *
 * @param {Function} func - The function to run.
 * @param {number} wait - Time in milliseconds.
 * @param {boolean} immediate - Whether we should run function immedately.
 * @returns {Function} the debounced function
 */
export function debounce(func, wait, immediate) {
  let timeout;
  return function () {
    const context = this,
      args = arguments;
    clearTimeout(timeout);
    timeout = setTimeout(function () {
      timeout = null;
      if (!immediate) func.apply(context, args);
    }, wait);
    if (immediate && !timeout) func.apply(context, args);
  };
}

/**
 * Makes sure that a URI has host, protocol, etc.
 *
 * @param {string} uri - the uri to test
 * @returns {string} a uri with the protocol, host and port if necessary
 */
export function getFullUrl(uri) {
  if (!uri) return uri;

  // https://gist.github.com/jlong/2428561
  const parser = document.createElement('a');
  parser.href = uri;
  return parser.href;
}

/**
 * processes error messages so we can see them in redux
 *
 * @param {Error} err - the error object from api
 * @returns {{name: string, message: string}} processed error object we can see
 */
export function processErrorMessage(err) {
  return {
    name: err.name,
    message: err.message,
  };
}

/**
 * Takes in a number and outputs to percentage
 *
 * @param {number} num - value we convert .9999
 * @returns {number} 99.99
 */
export function formatPercentage(num) {
  // we have to do this so it is a float and not a string
  const val = parseFloat(parseFloat(num * 100).toFixed(2));
  return isNaN(val) ? 0.0 : val;
}

/**
 * helper function
 *
 * @param {object} bucket - contains key value pairs
 * @returns {string} name of the key that has the buckets
 */
export const getSubKeyName = (bucket) => {
  for (const item in bucket) {
    if (item !== 'trend_period' && bucket[item].buckets) {
      return item;
    }
  }
  return '';
};

/**
 * helper function to take in array parameters from the url, filters, etc and
 * set the values in the processed object
 *
 * @param {object} params - the object from the URL_CHANGED action
 * @param {object} state - the state we will update with a single value or arr
 * @param {object} arrayParams - the array of strings that we will check against
 */
export const processUrlArrayParams = (params, state, arrayParams) => {
  arrayParams.forEach((field) => {
    if (typeof params[field] !== 'undefined') {
      if (typeof params[field] === 'string') {
        state[field] = [params[field]];
      } else {
        state[field] = params[field];
      }
    }
  });

  if (params.has_narrative) {
    state.has_narrative = !!params.has_narrative;
  } else {
    delete state.has_narrative;
  }
};

/**
 * gets a filter and its subagg filters
 *
 * @param {string} filterKey - the filter 'Debt'
 * @param {Array} subitems - the buckets to process to generate slug
 * @returns {Set<any>} returns a set of uniques Debt, Debt*Foo
 */
export const getAllFilters = (filterKey, subitems) => {
  const values = new Set();
  // Add the parent
  values.add(filterKey);
  // Add the shown subitems
  subitems.forEach((sub) => {
    values.add(slugify(filterKey, sub.key));
  });
  return values;
};

/**
 * Wrapper around analytics event action creator to minimize the copypasta
 *
 * @param {string} action - GA Action (not redux action)
 * @param {string} label - param used by GA
 */
export const sendAnalyticsEvent = (action, label) => {
  Analytics.sendEvent(Analytics.getDataLayerOptions(action, label));
};

/**
 * Helper function to get the selected class based on two different values
 *
 * @param {string|number|boolean} first - Value 1 to compare
 * @param {string|number|boolean} second - Value 2 to compare
 * @param {string} selectedClassName - The value that should be returned if both are the same
 * @returns {string} The selected class
 */
export const selectedClass = (
  first,
  second,
  selectedClassName = 'selected',
) => {
  return first === second ? ' ' + selectedClassName : '';
};

/**
 * Remove all properties with the value 'null' from the object, or empty string
 *
 * @param {object} object - the object with potential nulls
 * @returns {object} the processed object
 */
export function removeNullProperties(object) {
  const myObject = Object.keys(object).reduce((acc, key) => {
    if (
      object[key] !== null &&
      object[key] !== undefined &&
      object[key] !== '' &&
      !Number.isNaN(object[key])
    ) {
      acc[key] = object[key];
    }
    return acc;
  }, {});

  for (const key in myObject) {
    if (Array.isArray(myObject[key]) && myObject[key].length === 0) {
      delete myObject[key];
    }
  }

  return myObject;
}

/**
 * Builds a URL from a path and dictionary
 *
 * @param {string} path - The V2 endpoint.
 * @param {object} params - A key/value pair of the query string params.
 * @returns {string} The full endpoint url.
 */
export function formatUri(path, params) {
  return path + '?' + queryString.stringify(params);
}
