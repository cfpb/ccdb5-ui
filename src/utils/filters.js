import { filterPatch, SLUG_SEPARATOR } from '../constants';
import { slugify } from '../utils';

export const formatPillPrefix = (fieldName) => {
  // update this if they want the pill prefixes in other fields.
  if (fieldName === 'timely') {
    const rep = /_/g;
    const prefix = fieldName.replace(rep, ' ');
    return prefix[0].toUpperCase() + prefix.slice(1) + ': ';
  }
  return '';
};

export const getUpdatedFilters = (filterName, filters, aggs, fieldName) => {
  // early exit if its not issue or product
  if (!filterPatch.includes(fieldName)) {
    return filters;
  }

  // remove parent filter
  const parts = filterName.split(SLUG_SEPARATOR);
  const parentFilter = parts[0];
  const hasParent = filters.includes(parentFilter);
  // remove current filter
  const oldFilters = filters
    .filter((filter) => filter !== parentFilter && filterName)
    .filter((filter) => filter !== filterName);
  // apply siblings
  const sibs = [];
  if (hasParent) {
    const siblings = aggs.find((agg) => agg.key === parentFilter);
    siblings['sub_' + fieldName + '.raw'].buckets.forEach((bucket) => {
      // don't include self
      if (bucket.key !== parts[1]) {
        sibs.push(slugify(parentFilter, bucket.key));
      }
    });
  }

  return oldFilters.concat(sibs);
};
