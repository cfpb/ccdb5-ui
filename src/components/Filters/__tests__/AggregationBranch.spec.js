import React from 'react';
import { IntlProvider } from 'react-intl';
import { Provider } from 'react-redux';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import renderer from 'react-test-renderer';
import ReduxAggregationBranch, {
  AggregationBranch,
  CHECKED,
  mapDispatchToProps,
  UNCHECKED,
} from '../AggregationBranch';
import { slugify } from '../../../utils';

// ----------------------------------------------------------------------------
// Setup

const item = {
  key: 'foo',
  doc_count: 99,
};

const subitems = [
  { key: 'bar', doc_count: 90 },
  { key: 'baz', doc_count: 5 },
  { key: 'qaz', doc_count: 4 },
];

/**
 *
 * @param checkedState
 */
function setupEnzyme(checkedState = UNCHECKED) {
  const props = {
    activeChildren: [
      // not hidden
      slugify('foo', 'bar'),
      // hidden
      slugify('foo', 'quux'),
    ],
    checkedState,
    checkParent: jest.fn(),
    fieldName: 'issue',
    filters: [],
    item: item,
    subitems: subitems,
    uncheckParent: jest.fn(),
  };

  const target = shallow(<AggregationBranch {...props} />);

  return {
    props,
    target,
  };
}

/**
 *
 * @param selections
 */
function setupSnapshot(selections) {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    query: {
      issue: selections,
    },
  });

  return renderer.create(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ReduxAggregationBranch
          item={item}
          subitems={subitems}
          fieldName="issue"
        />
      </IntlProvider>
    </Provider>,
  );
}

// ----------------------------------------------------------------------------
// Test

describe('component::AggregationBranch', () => {
  describe('snapshots', () => {
    it('renders with all checked', () => {
      const selections = [
        'foo',
        slugify('foo', 'bar'),
        slugify('foo', 'baz'),
        slugify('foo', 'qaz'),
      ];
      const target = setupSnapshot(selections);
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders with indeterminate', () => {
      const selections = [slugify('foo', 'bar')];
      const target = setupSnapshot(selections);
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders with none checked', () => {
      const target = setupSnapshot([]);
      const tree = target.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('toggle behavior', () => {
    it('shows the children when the label is clicked', () => {
      const { target } = setupEnzyme();
      const theButton = target.find('button');

      expect(target.state('hasChildren')).toEqual(false);
      theButton.simulate('click');
      expect(target.state('hasChildren')).toEqual(true);
    });
  });

  describe('parent checkbox logic', () => {
    it('calls one action when the checkbox is already selected', () => {
      const { target, props } = setupEnzyme(CHECKED);
      const checkbox = target.find('li.parent input[type="checkbox"]');
      checkbox.simulate('change');
      expect(props.uncheckParent).toHaveBeenCalledWith('issue', [
        'foo',
        'foo•bar',
        'foo•baz',
        'foo•qaz',
        'foo•quux',
      ]);
      expect(props.checkParent).not.toHaveBeenCalled();
    });

    it('calls another action when the checkbox is not selected', () => {
      const { target, props } = setupEnzyme();
      const checkbox = target.find('li.parent input[type="checkbox"]');
      checkbox.simulate('change');
      expect(props.uncheckParent).not.toHaveBeenCalled();
      expect(props.checkParent).toHaveBeenCalledWith({
        fieldName: 'issue',
        filters: [],
        item: { doc_count: 99, key: 'foo' },
      });
    });
  });

  describe('mapDispatchToProps', () => {
    it('hooks into replaceFilters', () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).checkParent({
        fieldName: 'foo',
        filters: [
          slugify('bay', 'bee'),
          slugify('bay', 'ah'),
          'another filter',
        ],
        item: { key: 'bay' },
        values: ['bar', 'baz'],
      });
      expect(dispatch.mock.calls).toEqual([
        [
          {
            filterName: 'foo',
            requery: 'REQUERY_ALWAYS',
            type: 'FILTER_REPLACED',
            values: ['another filter', 'bay'],
          },
        ],
      ]);
    });

    it('hooks into removeMultipleFilters', () => {
      const dispatch = jest.fn();
      mapDispatchToProps(dispatch).uncheckParent({
        fieldName: 'foo',
        values: ['bar', 'baz'],
      });
      expect(dispatch.mock.calls.length).toEqual(1);
    });
  });
});
