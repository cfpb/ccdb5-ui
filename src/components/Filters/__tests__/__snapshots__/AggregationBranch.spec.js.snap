// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component::AggregationBranch snapshots renders with all checked 1`] = `
<div
  className="aggregation-branch foo"
>
  <li
    className="parent m-form-field m-form-field__checkbox body-copy"
  >
    <input
      aria-label="foo"
      checked={true}
      className="flex-fixed a-checkbox"
      id="issue-foo"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="toggle a-label"
      htmlFor="issue-foo"
    >
      <span
        className="u-visually-hidden"
      >
        foo
      </span>
    </label>
    <button
      className="flex-all a-btn a-btn__link"
      onClick={[Function]}
    >
      <span>
        foo
      </span>
      <svg
        className="cf-icon-svg"
        viewBox="0 0 988.7 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M923.6 967.6c-17.3 0-33.8-6.8-46-19L494.1 565 110.5 948.5c-25.6 25.1-66.8 24.8-91.9-.8-24.8-25.3-24.8-65.8 0-91.1l429.5-429.5c25.4-25.4 66.5-25.4 91.9 0l429.6 429.5c25.4 25.4 25.4 66.5.1 91.9-12.3 12.3-28.8 19.1-46.1 19.1z"
        />
      </svg>
    </button>
    <span
      className="flex-fixed parent-count"
    >
      <span>
        99
      </span>
    </span>
  </li>
  <ul
    className="children"
  >
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="foo•bar"
        checked={true}
        className="flex-fixed a-checkbox"
        id="issue-foo-bar"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="issue-foo-bar"
      >
        bar
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          90
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="foo•baz"
        checked={true}
        className="flex-fixed a-checkbox"
        id="issue-foo-baz"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="issue-foo-baz"
      >
        baz
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          5
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="foo•qaz"
        checked={true}
        className="flex-fixed a-checkbox"
        id="issue-foo-qaz"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="issue-foo-qaz"
      >
        qaz
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          4
        </span>
      </span>
    </li>
  </ul>
</div>
`;

exports[`component::AggregationBranch snapshots renders with indeterminate 1`] = `
<div
  className="aggregation-branch foo"
>
  <li
    className="parent m-form-field m-form-field__checkbox body-copy"
  >
    <input
      aria-label="foo"
      checked={false}
      className="flex-fixed a-checkbox"
      id="issue-foo"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="toggle a-label indeterminate"
      htmlFor="issue-foo"
    >
      <span
        className="u-visually-hidden"
      >
        foo
      </span>
    </label>
    <button
      className="flex-all a-btn a-btn__link"
      onClick={[Function]}
    >
      <span>
        foo
      </span>
      <svg
        className="cf-icon-svg"
        viewBox="0 0 988.7 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M923.6 967.6c-17.3 0-33.8-6.8-46-19L494.1 565 110.5 948.5c-25.6 25.1-66.8 24.8-91.9-.8-24.8-25.3-24.8-65.8 0-91.1l429.5-429.5c25.4-25.4 66.5-25.4 91.9 0l429.6 429.5c25.4 25.4 25.4 66.5.1 91.9-12.3 12.3-28.8 19.1-46.1 19.1z"
        />
      </svg>
    </button>
    <span
      className="flex-fixed parent-count"
    >
      <span>
        99
      </span>
    </span>
  </li>
  <ul
    className="children"
  >
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="foo•bar"
        checked={true}
        className="flex-fixed a-checkbox"
        id="issue-foo-bar"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="issue-foo-bar"
      >
        bar
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          90
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="foo•baz"
        checked={false}
        className="flex-fixed a-checkbox"
        id="issue-foo-baz"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="issue-foo-baz"
      >
        baz
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          5
        </span>
      </span>
    </li>
    <li
      className="layout-row m-form-field m-form-field__checkbox"
    >
      <input
        aria-label="foo•qaz"
        checked={false}
        className="flex-fixed a-checkbox"
        id="issue-foo-qaz"
        onChange={[Function]}
        type="checkbox"
      />
      <label
        className="a-label flex-all bucket-key body-copy"
        htmlFor="issue-foo-qaz"
      >
        qaz
      </label>
      <span
        className="flex-fixed bucket-count"
      >
        <span>
          4
        </span>
      </span>
    </li>
  </ul>
</div>
`;

exports[`component::AggregationBranch snapshots renders with none checked 1`] = `
<div
  className="aggregation-branch foo"
>
  <li
    className="parent m-form-field m-form-field__checkbox body-copy"
  >
    <input
      aria-label="foo"
      checked={false}
      className="flex-fixed a-checkbox"
      id="issue-foo"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="toggle a-label"
      htmlFor="issue-foo"
    >
      <span
        className="u-visually-hidden"
      >
        foo
      </span>
    </label>
    <button
      className="flex-all a-btn a-btn__link"
      onClick={[Function]}
    >
      <span>
        foo
      </span>
      <svg
        className="cf-icon-svg"
        viewBox="0 0 988.2 1200"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M494.1 967.2c-17.3 0-33.8-6.8-46-19L18.6 518.6c-25.1-25.6-24.8-66.8.8-91.9 25.3-24.8 65.8-24.8 91.1 0l383.6 383.6 383.6-383.6c25.6-25.1 66.8-24.8 91.9.8 24.8 25.3 24.8 65.8 0 91.1L540.1 948.1c-12.2 12.2-28.7 19.1-46 19.1z"
        />
      </svg>
    </button>
    <span
      className="flex-fixed parent-count"
    >
      <span>
        99
      </span>
    </span>
  </li>
</div>
`;
